pkg/xds/server/callbacks/dataplane_lifecycle.go

```diff
@@ -106,26 +113,51 @@ func (d *DataplaneLifecycle) OnProxyDisconnected(streamID core_xds.StreamID, dpK
 
 var _ DataplaneCallbacks = &DataplaneLifecycle{}
 
-func NewDataplaneLifecycle(appCtx context.Context, resManager manager.ResourceManager) *DataplaneLifecycle {
+func NewDataplaneLifecycle(appCtx context.Context, resManager manager.ResourceManager, authenticator xds_auth.Authenticator) *DataplaneLifecycle {
 	return &DataplaneLifecycle{
 		resManager:           resManager,
+		authenticator:        authenticator,
 		createdDpByCallbacks: map[model.ResourceKey]mesh_proto.ProxyType{},
 		appCtx:               appCtx,
 	}
 }
 
-func (d *DataplaneLifecycle) registerDataplane(dp *core_mesh.DataplaneResource) error {
+func (d *DataplaneLifecycle) registerDataplane(ctx context.Context, dp *core_mesh.DataplaneResource) error {
 	key := model.MetaToResourceKey(dp.GetMeta())
 	existing := core_mesh.NewDataplaneResource()
 	return manager.Upsert(d.resManager, key, existing, func(resource model.Resource) error {
+		if err := d.validateUpsert(ctx, existing); err != nil {
+			return errors.Wrap(err, "you are trying to override existing dataplane to which you don't have an access.")
+		}
 		return existing.SetSpec(dp.GetSpec())
 	})
 }
 
-func (d *DataplaneLifecycle) registerZoneIngress(zi *core_mesh.ZoneIngressResource) error {
+// validateUpsert checks if a new data plane proxy can replace the old one.
+// We cannot just upsert a new data plane proxy over the old one, because if you generate token bound to mesh + kuma.io/service
+// then you would be able to just replace any other data plane proxy in the system.
+// Ideally, when starting a new data plane proxy, the old one should be deleted, so we could do Create instead of Upsert, but this may not be the case.
+// For example, if you spin down CP, then DP, then start CP, the old DP is still there for a couple of minutes (see pkg/gc).
+// We could check if Dataplane is identical, but CP may alter Dataplane resource after is connected.
+// What we do instead is that we use current data plane proxy credential to check if we can manage already registered Dataplane.
+func (d *DataplaneLifecycle) validateUpsert(ctx context.Context, existing model.Resource) error {
+	if util_proto.IsEmpty(existing.GetSpec()) { // existing DP is empty, resource does not exist
+		return nil
+	}
+	credential, err := xds_auth.ExtractCredential(ctx)
+	if err != nil {
+		return err
+	}
+	return d.authenticator.Authenticate(ctx, existing, credential)
```

Comment:
This api is strange to extract the creds from the context to pass down what was extracted along with this context to the authenticator. Any reason why ExtractCredential isn't called inside the authenticator?

Reply:
Looking at the code it seems like this would force people to use the correct context in hds and xds auth.
Though this can be done as a follow up issue/PR